homeassistant/components/generic/config_flow.py

```diff
@@ -309,6 +316,40 @@ def register_preview(hass: HomeAssistant) -> None:
     hass.data[DOMAIN][IMAGE_PREVIEWS_ACTIVE] = True
 
 
+async def register_stream_preview(hass: HomeAssistant, config) -> str:
+    """Set up preview for camera stream during config flow."""
+    hass.data.setdefault("camera", {})
+
+    # Need to load the camera prefs early to avoid errors generating the stream
+    # if the user does not already have the stream component loaded.
+    if hass.data.get(DATA_CAMERA_PREFS) is None:
+        prefs = CameraPreferences(hass)
+        await prefs.async_load()
+        hass.data[DATA_CAMERA_PREFS] = prefs
+
+    # Create a camera but don't add it to the hass object.
+    cam = GenericCamera(hass, config, "stream_preview", "Camera Preview Stream")
+    cam.entity_id = DOMAIN + ".stream_preview"
+    cam.platform = EntityPlatform(
+        hass=hass,
+        logger=_LOGGER,
+        domain=DOMAIN,
+        platform_name="camera",
+        platform=None,
+        scan_interval=timedelta(seconds=1),
+        entity_namespace=None,
+    )
+
+    stream = await cam.async_create_stream()
```

Comment:
Perhaps `camera._async_stream_endpoint_url` should be exposed for this purpose.

Reply:
I just re-read how `async_test_stream` works and that seems to me to argue for a direct stream integration since one already exists.
- Whats the motivation for multiple ways to do that? 
- Whats going to happen to the image preview?  seems like its a parallel path, but can it be combined?

Reply:
Yes, I agree on the `camera._async_stream_endpoint_url`, exposing this would avoid some duplication.

In general my starting approach was to simulate the creation of the final camera entity as much as possible, with all of its settings, so that the user gets real information on whether that configuration will actually work. This seemed the most reliable & future proof way of doing it. So I see the preview (still and stream) equally about testing the behaviour of the entity that will be created as showing an image/stream.

There are two phases:
1. Verify that the settings are valid
This happens when the user clicks submit on the first config flow screen.

2. Show a preview
Create something with those settings and show it. This happens when the websocket callback is triggered on the second screen.

The issue I found was that during 1) I'm not ready to create the preview. The settings could be wrong, in which case I need to bounce the user back to the first screen with an error next to the wrong parameter.

But via the websocket method, the notice to generate the preview only comes when we are already at step 2). If I find an error at that step, I need to bounce them back to step 1) again, but I couldn't see a way of doing that since the step 2) dialog is already showing.

So we end up with two hits to the external URLs, one for checking and one for the preview. Not a big issue in the case of the still image, but 2x the stream delay is very noticeable.  I don't really like this and I'm open to ideas for a better approach. Possibly I could create the stream during the check and keep hold of it if the check was successful.


Regarding the image preview, yes you are right, I was planning to integrate that back into the same approach in a separate PR.

Side note: I did try for a while with a live preview at the bottom of the settings screen in the same way that `time_date` does it. But I stopped down that route because:
1. the dialog became incredibly tall, so you couldn't actually see the preview without scrolling
2. the external URLs were being hit and streams generated for every character typed, which felt likely to cause problems (e.g. authentication lockouts)

There might be a potential architecture improvement here, because I imagine most integrations will face similar issues, and will want to run a preview without interfering with the real hass object (also during options flow).

Reply:
@allenporter Ok, I've progressed it quite a bit further now. The stream test and the preview are now combined, this makes it a lot faster too.

Left over streams are now auto-stopped after 10minutes

I've tidied up the rest of the code and removed several things that were unnecessary.

Since I've 'sort of' subclassed Stream anyway now to do the timeout, that could possibly open up the option for blocking `async_write_ha_state`.

Either way I would like your thoughts on this before going further. Thanks in advance!

Reply:
I really like this approach a lot and it seems very simple and easy to read/understand. I really like that we don't need to use the entity anymore here and can just have one way to use the stream.

Any other specific input now or review needed? It generally looks good and i wanted to give quick feedback, but happy to do the last round of review when you are ready.

(The part where it loads the platform to ensure translations are loaded is something i'll want to give a closer look at for prior art/ alternatives)

Reply:
Thanks @allenporter  I'll get the tests implemented, then go for final review.

I will put the options flow on a separate PR.

Reply:
I may have missed before that the image preview works by using the camera entity, so now the image preview and stream preview take different strategies.